walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,64 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    ArrayList<TimeRange> result = new ArrayList<>();
+    List<TimePoint> timePoints = new ArrayList<>();
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> allAttendees = new HashSet<>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+    allAttendees.addAll(attendees);
+    result = queryByAttendenceAndDuration(events, allAttendees, duration);
+    if (result.size() == 0) {
+      result = queryByAttendenceAndDuration(events, attendees, duration);
+    }
+    return result;
+  }
+
+  public ArrayList<TimeRange> queryByAttendenceAndDuration(
+      Collection<Event> events, Set<String> attendees, long duration) {
+    ArrayList<TimeRange> result = new ArrayList<>();
+    List<TimePoint> timePoints = new ArrayList<>();
+    for (Event event : events) {
+      Set<String> curAttendees = new HashSet<>(event.getAttendees());
+      curAttendees.retainAll(attendees);
+      if (!curAttendees.isEmpty()) {
```

Comment:
This works :) FYI: You could also have used `if(!Collections.disjoint(event.getAttendees(), attendees))`